### ISSUE TITLE
Wandering Monster redirects attacks that can no longer be carried out, and other Secrets can react to this

### ISSUE THREAD

**Issue by Jetz72** [HearthSim/hs-bugs#1098]
https://www.reddit.com/r/hearthstone/comments/aglhb7/why_does_freezing_trap_wandering_monster_snake/
>why does freezing trap -> wandering monster -> snake trap trigger all 3 traps?
>just had this happen in a game. it's weird enough that a wandering monster triggers after freezing trap, but why does a snake trap trigger after that? my minion was already in my hand before wandering monster triggered, it never attacked the 3 drop at all. this just doesn't make sense to me I guess
>edit: I get why freezing -> wandering happens, but the snake trap really shouldn't trigger right? no minion was attacked, even from the secrets there was no minion attacked

Tested and confirmed:
![Screenshot 1](https://i.imgur.com/BFeckNF.png)
![Screenshot 2](https://i.imgur.com/zeYFUam.png)

While it's natural that Wandering Monster triggers and summons the 3-drop, it's fairly counterintuitive that the attack can be meaningfully redirected after the minion is no longer in play.

Seems worth testing whether it's possible to have a minion leave and reenter the battlefield over the span of an attack and still carry the same attack out.

**Comment by Kybxd** [comment]
A possible plan: have your demon attack enemy hero and Freeze Trap returns it to hand. Then Explosive Trap → Acolyte of Pain draws Voidcaller → Dollmaster → Spiritsinger resummons the demon.

**Comment by Jetz72** [comment]
The one I suggested was Freezing Trap followed by Wandering Monster, 3-drop triggers a Knife Juggler, Juggler kills a Piloted Reaper to bring it back out. @troggnostupidhs tested that but looking over the screenshots, it seems the setup was missing something to force a death phase and actually trigger the deathrattle before the attack ended. An Acolyte and a Scroll of Wonder in the deck would be able to do that, though your setup seems more reliable overall.

**Comment by Kybxd** [comment]
The death phase between Proposed Attack Phase and Attack Phase is cancelled. I'm not sure when the change occurs but we found it after patch 13.0. Here is how Combat Sequence resolves for now:

1. Combat Phase
(1) Proposed Attack triggers queue and resolve.
(2) If defender changes, repeat (1) for the new defender. This repeats until defender doesn't change.
(3) Attack triggers queue and resolve.
(4) If attacker, defender or either hero is mortally wounded/pending destroy, or has left board, the Combat is cancelled. (5) and (6) are skipped.
(5) Combat resolves.
(6) After Attack triggers queue and resolve.
2. Runespear Phase
If The Runespear discovers a spell in the previous phase, the spell is resolved here.

Since we have Spiritsinger/Dollmaster, it's possible to summon a specific minion, or trigger a specific deathrattle whenever we want. Here is an interesting video showing that your weapon losing durability is an After Attack trigger. The order between it and enemy Bear Trap is decided by Dominant Player. https://www.bilibili.com/video/av40452958 (4:50 to 6:30)

**Comment by Jetz72** [comment]
Yeah, there's no natural death phase in there. But you can often force one at any time by using Acolyte of Pain to draw a Scroll of Wonder, which will force the death phase. Or are you saying that forced death phases at that timing will be cancelled as well? Either way, the Dollmaster+Spiritsinger combo seems much more practical.

Is the Runespear really considered its own phase? Could it not just have its own forced death phase before it casts the spell? Or is there more to it than that?

**Comment by PattuX** [comment]
Also works when you played Snake Trap before Wandering Monster: https://www.youtube.com/watch?v=4c7OPWMgjy8

This indicates the attack sequence is only interrupted when the attacker is not in play before step (3). It doesn't matter whether it is in play when you go back from (2) to (1). If that's really the case, then the setup bouncing the attacker to hand and re-summoning it should work.

**Comment by troggnostupidhs** [comment]
> Setup: Player A (Hunter) has Freezing Trap and Explosive trap, in that order. Player B has an empty deck. B has a 0-cost minion (ideally with health buffed to 3 or more but it probably doesn't matter), Piloted Reaper, Acolyte of Pain, Umbra, and Dorian in play. B plays Baleful Banker on the Reaper, making it the only card in the deck. B attacks the face with the 0-drop. Freezing Trap sends the minion back to the hand. Explosive Trap damages the Acolyte, Acolyte draws the Reaper, Dorian summons a Reaper, Umbra triggers the Reaper, Reaper brings the 0 (now 2)- drop back into play, and possibly the attack completes.

![hearthstone screenshot 01-22-19 20 26 45](https://user-images.githubusercontent.com/22508641/51576436-3417c480-1e84-11e9-9f0f-293e70cef5b2.png)

![hearthstone screenshot 01-22-19 20 28 25](https://user-images.githubusercontent.com/22508641/51576656-1e56cf00-1e85-11e9-8aca-d3fa86e4472f.png)

note: 1 damage was probably fatigue

Hunter
AAEBAR8AD40B+wGcAqgChwTJBO0GgAeXCMUIhsMCmM0CnM0C4eMCpvACAA==

Priest
AAEBAa0GAsnCAsLxAg6zAfsBlwKcAuUEyQbXCvsMt2ziygLwzwKm8AKn9wLf+AIA

**Comment by Kybxd** [comment]
@Jetz72 Only the natural death phase is cancelled. I classify the Runespear as a phase because if there is  another After Attack trigger which is later than the Runespear (for example enemy Secondary Player's Bear Trap), the order is: the Runespear discovers a spell - Bear Trap summons a bear - death phase - the spell resolves and it may affect the bear.

**Comment by Jetz72** [comment]
Regarding Trogg's test, the log here - https://gist.github.com/troggnostupidhs/2e6a3d0764991b035563a9e304a7516a - shows a couple points relevant to this.

First, when the Freezing Trap resolves, before the attacker returns to the hand, it is given this tag:
```
D 20:27:01.6367237 PowerTaskList.DebugPrintPower() - TAG_CHANGE Entity=[entityName=Wisp id=36 zone=PLAY zonePos=3 cardId=CS2_231 player=2] tag=SHOULDEXITCOMBAT value=1 
```

The name of that tag alone is a hint at what the intended behavior is. But apparently, it does not get applied until after Explosive Trap - the last secret that triggered off the attack:
```
D 20:27:17.0217121 PowerTaskList.DebugDump() - Block Start=(null)
D 20:27:17.0217121 PowerTaskList.DebugPrintPower() -     TAG_CHANGE Entity=[entityName=Explosive Trap id=18 zone=SECRET zonePos=0 cardId=EX1_610 player=1] tag=1068 value=4 
D 20:27:17.0217121 PowerTaskList.DebugPrintPower() -     TAG_CHANGE Entity=[entityName=Explosive Trap id=18 zone=SECRET zonePos=0 cardId=EX1_610 player=1] tag=1068 value=0 
D 20:27:17.0217121 PowerTaskList.DebugPrintPower() -     TAG_CHANGE Entity=[entityName=Explosive Trap id=18 zone=SECRET zonePos=0 cardId=EX1_610 player=1] tag=ZONE value=GRAVEYARD 
D 20:27:17.0217121 PowerTaskList.DebugPrintPower() -     TAG_CHANGE Entity=[entityName=Wisp id=36 zone=PLAY zonePos=7 cardId=CS2_231 player=2] tag=NUM_ATTACKS_THIS_TURN value=1 
D 20:27:17.0217121 PowerTaskList.DebugPrintPower() -     TAG_CHANGE Entity=GameEntity tag=PROPOSED_ATTACKER value=0 
D 20:27:17.0217121 PowerTaskList.DebugPrintPower() -     TAG_CHANGE Entity=GameEntity tag=PROPOSED_DEFENDER value=0 
D 20:27:17.0217121 PowerTaskList.DebugPrintPower() -     TAG_CHANGE Entity=[entityName=Wisp id=36 zone=PLAY zonePos=7 cardId=CS2_231 player=2] tag=SHOULDEXITCOMBAT value=0 
D 20:27:17.0217121 PowerTaskList.DebugPrintPower() - BLOCK_END
```

So Freezing Trap does cancel combat, but not between the secrets. The attack is still ongoing, and Wandering Monster doesn't care about whether combat should continue when it updates the PROPOSED_DEFENDER in the problem case. Snake Trap (or Venomstrike Trap) sees it the same way it sees "attack a minion" -> freezing -> snakes. Combat is only properly cancelled after all the secrets trigger.
